[6.1] Subformfield groupByFieldset handling - Update default.php

BrainforgeUK opened the pull request:
groupByFieldset functionality added.

### Summary of Changes

Setting _groupByFieldset_ parameter in XML to _true_ does not currently serve any purpose.

This change (without any other fieldset parameters) uses CSS grid display to group fields in a fieldset.

Additional parameters can be used with the fieldset(s) can be used to customise how the fields are grouped (e.g. use flex instead of grid).

### Testing Instructions

Install the following plugins:
[bfsubfieldtest1.zip](https://github.com/user-attachments/files/16116194/bfsubfieldtest1.zip)
[bfsubfieldtest2.zip](https://github.com/user-attachments/files/16116195/bfsubfieldtest2.zip)
[bfsubfieldtest3.zip](https://github.com/user-attachments/files/16116196/bfsubfieldtest3.zip)

### Actual result BEFORE applying this Pull Request

Go to plugin admin page.
Unexpected result - there is no difference between _bfsubfieldtest1_ and _bfsubfieldtest2_.

### Expected result AFTER applying this Pull Request

Go to plugin admin page.
With _bfsubfieldtest2_ the  the fields are now grouped by fieldset.
_bfsubfieldtest3_ illustrates how the appearance of fields grouped within fieldsets can be customised.

### Link to documentations
The code in this pull request includes comments about the available fieldset attributes. These comments together with the example subform fields from _bfsubfieldtest3_ need to be included in the documentation.
Please select:
- [*] Documentation link for docs.joomla.org:
[Avanced Form Guide](https://docs.joomla.org/Advanced_form_guidel)
[Subform Form Field](https://docs.joomla.org/Subform_form_field_type)
- [] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

BrainforgeUK committed:
Update default.php

groupByFieldset functionality added.

richard67 commented:
@BrainforgeUK Please fix the PHP code style errors reported here: https://ci.joomla.org/joomla/joomla-cms/77074/1/7

BrainforgeUK committed:
Update default.php

Corrected.

richard67 commented on layouts/joomla/form/field/subform/default.php:
```diff
@@ -1,11 +1,10 @@
- *
- * @copyright   (C) 2016 Open Source Matters, Inc. <https://www.joomla.org>
- * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ * @package   Component for customer to book slot appointments with site owner.
+ * @version   0.0.1
+ * @author    https://www.brainforge.co.uk
+ * @copyright Copyright (C) 2024 Jonathan Brain. All rights reserved.
+ * @license   GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
```
I think this change is wrong and should be reverted.

BrainforgeUK replied:
Ok the original was a copy and paste from a component ... which was done badly!
Looks better now, see what the checks throw up.

BrainforgeUK committed:
Update default.php

Corrected.

BrainforgeUK committed:
Update default.php

Corrected space: foreach (...)

BrainforgeUK committed:
Update default.php

Corrected space.

Fedik commented:
Thanks for the PR, a few notes:

- Please remove iinline style, CSS in XML will not be accepted. You can use only class.
- Please use `fireach(): ... endforeach;` and `if (): ... endif` in the code that mixing PHP and HTML.

Fedik commented on layouts/joomla/form/field/subform/default.php:
```diff
@@ -35,7 +35,61 @@
+                <?php
+                foreach ($fields as $field) {
+                    ?>
+                    <div class="<?php echo $fieldSet->name . '-fieldwrapper ' .
```
This need to remove, the field already rendered in its own wraper. https://github.com/joomla/joomla-cms/blob/e7f5cc182bb1cd39c7f085d529888ac4fa6fc77e/layouts/joomla/form/renderfield.php#L47

BrainforgeUK replied:
- I've changed the if / foreach styling.
- Now pass field wrapping classes into renderField().
- Added missing explanation of 'fieldwrapperstyle' fieldset attribute.

Corrected test plugin:
[bfsubfieldtest3.zip](https://github.com/user-attachments/files/16142799/bfsubfieldtest3.zip)

Added another test plugin to demonstrate 'fieldwrapperstyle' fieldset attribute.
[bfsubfieldtest4.zip](https://github.com/user-attachments/files/16142804/bfsubfieldtest4.zip)

The inline styling works in the XML as shown in the test plugins.

BrainforgeUK committed:
Update default.php

Now uses foreach(): ... endforeach; and if (): ... endif coding style.

BrainforgeUK committed:
Update default.php

Passed field wrapping classes to renderFied(). Also added comment to describe 'fieldwrapperstyle' fieldset attribute.

BrainforgeUK committed:
Update default.php

Changed indentation.

Quy commented on layouts/joomla/form/field/subform/default.php:
```diff
@@ -35,7 +35,60 @@
+        <?php foreach ($form->getGroup('') as $field) : ?>
+            <?php echo $field->renderField(); ?>
+        <?php endforeach; ?>
+    <?php endif ?>
```
```suggestion
    <?php endif; ?>
```

BrainforgeUK replied:
I take it you committed this suggestion, thanks.

brianteeman replied:
no it is for you to commit

QuyTon committed:
Update layouts/joomla/form/field/subform/default.php

BrainforgeUK commented:
Phan failing - is that something for me to address?

BrainforgeUK committed:
Update default.php

Quy commented on layouts/joomla/form/field/subform/default.php:
```diff
@@ -35,7 +35,60 @@
+             *                      Default: "1rem"
+             *  style               Optional style properties for this field set.
+             *                      Example: "border:1px dotted #ccc;padding:2px;box-shadow:2px 2px #888;"
+             *  fieldwrapperstyle   Optional style properties for wrapper around each field in thie field set.
```
```suggestion
             *  fieldwrapperstyle   Optional style properties for wrapper around each field in this field set.
```

Quy commented on layouts/joomla/form/field/subform/default.php:
```diff
@@ -35,7 +35,60 @@
+             */
+
+            $fields = $form->getFieldset($fieldSet->name);
+
```
Remove blank lines and align `=`.

Quy commented on layouts/joomla/form/field/subform/default.php:
```diff
@@ -35,7 +35,60 @@
+            <div class="<?php echo $fieldSet->class ?? ''; ?>"
+                 style="<?php echo implode('', $fieldsetStyle); ?>">
+                <?php foreach ($fields as $field) : ?>
+                    <div style="<?php echo $fieldSet->fieldwrapperstyle ?? ''; ?>">
```
Preferably to not display `style=""` when empty.

brianteeman commented:
we really should not be using inline styles

richard67 commented:
> Phan failing - is that something for me to address?

@BrainforgeUK No. Phan is expected to fail.

But you should check the other reviews by others. And you shouldn’t use inline styles as that will not allow to use a strict CSP.

BrainforgeUK committed:
Update layouts/joomla/form/field/subform/default.php

Co-authored-by: Quy <quy@nomonkeybiz.com>

BrainforgeUK committed:
Update default.php

BrainforgeUK committed:
Update default.php

Restructured class and style div element options.

Hackwar committed:
Merge branch '5.2-dev' into patch-27

mabdelaziz77 commented:
I have tested this item :white_check_mark: successfully on 9a598f2d8dd9d3c27aacfb8983bdde63dcf3b8e7<br /><br />I can see the difference between the two plugin.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43747">issues.joomla.org/tracker/joomla-cms/43747</a>.</sub>

degobbis commented on layouts/joomla/form/field/subform/default.php:
```diff
@@ -35,7 +35,74 @@
+            }
+
+            ?>
+            <div<?php echo $fieldSetOptions; ?>>
```
As “groupedByFieldsets”, wouldn't it also be logical to use the `<fieldset>` tag instead of `<div>`?

dautrich commented:
I have tested this item :white_check_mark: successfully on 9a598f2d8dd9d3c27aacfb8983bdde63dcf3b8e7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43747">issues.joomla.org/tracker/joomla-cms/43747</a>.</sub>

richard67 commented:
Not setting RTC despite of the 2 successful human tests as this PR adds inline styles which will result in CSP violations when using a reasonably strict CSP rule.

HLeithner commented:
This pull request has been automatically rebased to 5.3-dev.

HLeithner commented:
This pull request has been automatically rebased to 6.0-dev.

richard67 committed:
Merge branch '6.0-dev' into patch-27
